File: taurus-protect-sdk-python/taurus_protect/helpers/price_verifier.py
"""Price signature verification.

Rate and decimals feed amount conversion, so an unverified price is a wrong number a
caller acts on.
"""

from __future__ import annotations

import json
from typing import TYPE_CHECKING, List, Optional

from taurus_protect.crypto.signing import verify_signature
from taurus_protect.errors import IntegrityError

if TYPE_CHECKING:
    from taurus_protect.models.governance_rules import DecodedRulesContainer
    from taurus_protect.models.statistics import Price


def price_signed_bytes(price: "Price") -> bytes:
    """Return the exact bytes a price signature covers.

    This is validatord's CurrencyPrice JSON projection: the five fields it serialises, in
    its field order, with no spaces. The key order IS the signed byte sequence — do not
    reorder or add fields.
    """
    canonical = {
        "blockchain": price.blockchain or "",
        "currencyFrom": price.currency_from or "",
        "currencyTo": price.currency_to or "",
        "decimals": price.decimals or "",
        "rate": price.rate or "",
    }
    return json.dumps(canonical, separators=(",", ":")).encode("utf-8")


def verify_price(price: "Price", rules_container: "DecodedRulesContainer") -> None:
    """Check a price against the PRICEUPDATER keys in a verified rules container.

    The container decides whether prices must be signed at all. It is SuperAdmin-verified,
    so "this tenant has no price signer" is trustworthy, whereas "this price carries no
    signatures" is not — which is why a stripped signatures list on a tenant that DOES
    have a PRICEUPDATER is an error rather than a skip.

    Raises:
        IntegrityError: If no PRICEUPDATER signature verifies.
    """
    if price is None:
        raise IntegrityError("price cannot be None")
    if rules_container is None:
        raise IntegrityError("rules container required for price signature verification")

    keys = _price_updater_keys(rules_container)
    if not keys:
        # This tenant does not sign prices.
        return

    if not price.signatures:
        raise IntegrityError(
            f"price {price.currency_from}/{price.currency_to} carries no signatures but "
            "the rules container configures a PRICEUPDATER"
        )

    data = price_signed_bytes(price)
    for sig in price.signatures:
        if not sig.signature:
            continue
        for key in keys:
            try:
                if verify_signature(key, data, sig.signature):
                    return
            except Exception:  # noqa: BLE001 - a malformed signature is just a non-match
                continue

    raise IntegrityError(
        f"no PRICEUPDATER signature verifies for price "
        f"{price.currency_from}/{price.currency_to} "
        f"({len(price.signatures)} signature(s) offered)"
    )


def verify_prices(
    prices: List["Price"], rules_container: "DecodedRulesContainer"
) -> None:
    """Verify each price, raising on the first failure."""
    for price in prices or []:
        verify_price(price, rules_container)


def _price_updater_keys(rules_container: "DecodedRulesContainer") -> List[object]:
    """Public keys of every user carrying the PRICEUPDATER role."""
    keys: List[object] = []
    for user in getattr(rules_container, "users", None) or []:
        roles = getattr(user, "roles", None) or []
        if "PRICEUPDATER" not in roles:
            continue
        pem = getattr(user, "public_key_pem", None)
        if not pem:
            continue
        try:
            keys.append(rules_container.get_user_public_key(pem))
        except ValueError:
            continue
    return keys

File: taurus-protect-sdk-typescript/src/helpers/price-verifier.ts
/**
 * Price signature verification.
 *
 * Rate and decimals feed amount conversion, so an unverified price is a wrong number a
 * caller acts on.
 */

import type { KeyObject } from "crypto";

import { verifySignature, decodePublicKeyPem } from "../crypto";
import { IntegrityError } from "../errors";
import type { DecodedRulesContainer } from "../models/governance-rules";
import type { Price } from "../models/price";

/**
 * Returns the exact bytes a price signature covers.
 *
 * This is validatord's CurrencyPrice JSON projection: the five fields it serialises, in
 * its field order, with no spaces. The key order IS the signed byte sequence — do not
 * reorder or add fields.
 */
export function priceSignedBytes(price: Price): Buffer {
  const canonical =
    `{"blockchain":${JSON.stringify(price.blockchain ?? "")},` +
    `"currencyFrom":${JSON.stringify(price.currencyFrom ?? "")},` +
    `"currencyTo":${JSON.stringify(price.currencyTo ?? "")},` +
    `"decimals":${JSON.stringify(price.decimals ?? "")},` +
    `"rate":${JSON.stringify(price.rate ?? "")}}`;
  return Buffer.from(canonical, "utf-8");
}

/** Public keys of every user carrying the PRICEUPDATER role. */
function priceUpdaterKeys(rulesContainer: DecodedRulesContainer): KeyObject[] {
  const keys: KeyObject[] = [];
  for (const user of rulesContainer.users ?? []) {
    if (!user.roles?.includes("PRICEUPDATER") || !user.publicKeyPem) {
      continue;
    }
    try {
      keys.push(decodePublicKeyPem(user.publicKeyPem));
    } catch {
      // A key this SDK cannot decode cannot verify anything.
    }
  }
  return keys;
}

/**
 * Checks a price against the PRICEUPDATER keys in a verified rules container.
 *
 * The container decides whether prices must be signed at all. It is SuperAdmin-verified,
 * so "this tenant has no price signer" is trustworthy, whereas "this price carries no
 * signatures" is not — which is why a stripped signatures array on a tenant that DOES
 * have a PRICEUPDATER is an error rather than a skip.
 *
 * @throws {@link IntegrityError} If no PRICEUPDATER signature verifies
 */
export function verifyPrice(
  price: Price,
  rulesContainer: DecodedRulesContainer
): void {
  if (!price) {
    throw new IntegrityError("price cannot be null or undefined");
  }
  if (!rulesContainer) {
    throw new IntegrityError(
      "rules container required for price signature verification"
    );
  }

  const keys = priceUpdaterKeys(rulesContainer);
  if (keys.length === 0) {
    // This tenant does not sign prices.
    return;
  }

  const signatures = price.signatures ?? [];
  if (signatures.length === 0) {
    throw new IntegrityError(
      `price ${price.currencyFrom}/${price.currencyTo} carries no signatures but the ` +
        `rules container configures a PRICEUPDATER`
    );
  }

  const data = priceSignedBytes(price);
  for (const sig of signatures) {
    if (!sig.signature) {
      continue;
    }
    for (const key of keys) {
      try {
        if (verifySignature(key, data, sig.signature)) {
          return;
        }
      } catch {
        // A malformed signature is simply not a match.
      }
    }
  }

  throw new IntegrityError(
    `no PRICEUPDATER signature verifies for price ` +
      `${price.currencyFrom}/${price.currencyTo} ` +
      `(${signatures.length} signature(s) offered)`
  );
}

/** Verifies each price, throwing on the first failure. */
export function verifyPrices(
  prices: Price[],
  rulesContainer: DecodedRulesContainer
): void {
  for (const price of prices ?? []) {
    verifyPrice(price, rulesContainer);
  }
}
